### DISCUSSION TITLE
Console.WriteLine not showing up from AspNetCore http handler

### DISCUSSION THREAD

**Discussion by mxschmitt** [nunit/nunit#5067] [How to]
Repro:

1. `git clone https://github.com/mxschmitt/dotnet-nunit-asp-console-writeline`
2. `cd dotnet-nunit-asp-console-writeline`
3. `dotnet test --logger:"console;verbosity=detailed"`

Expected: All the 8 `WriteLine` calls are getting shown

Actual: they are not getting shown

Notes: 

- It looks like that the ExecutionContext is getting lost
- I have a "hack" inside which makes it work and demonstrates the bug, maybe its a known limitation:

https://github.com/mxschmitt/dotnet-nunit-asp-console-writeline/blob/a889dae8286c6fd0fc8383668a498303e9a1beeb/UnitTest1.cs#L24-L28

It seems related to https://github.com/nunit/nunit/issues/4728 - when I apply the same fix to the repro of https://github.com/nunit/nunit/issues/4728  its also working as expected.

Feel free to move / fold it into another issue. I filed it mostly for the sake of sharing the research.

**Comment by OsirisTerje** [comment]
@CharliePoole @manfred-brands It looks a bit like this should have been set internally.  Any thoughts ?

PS,  For a normal  ` await Task.Run(() =>`  with the same output statements, all works.

**Comment by mxschmitt** [comment]
Might be worth adding: a hack like this would work as well:

```csharp
var capturedExecutionContext = ExecutionContext.Capture();
app.Run(async context =>
{
    ExecutionContext.Restore(capturedExecutionContext);
```

Not sure if it would break any AspNetCore internals.

**Comment by CharliePoole** [comment]
Your probably right... with the usual proviso that I haven't worked seriously on the framework since 2017 and anything I say could be out of date. But maybe history helps a bit. :-) 

`TestExecutionContext` is one of those historically exposed classes in our Internal namespace that we were once trying to get rid of by either making them truly internal. That would have meant exposing more functionality in TestContext itself.
But that never happened, so here we are.

As things stand now, I think it should be made to work. It might help to get a non-Asp example to work with but if not the example @mxschmitt provided will work as a test if you remove his fix.

It would be helpful to know if the text output that is not displayed is getting into the TestResult.xml file at all.

Sorry I can't be more helpful, but it's nice that you ask. :-)

Fun fact, BTW... The top three all-time contributors to the have contributed either only spottily or not at all since 2020. In my case, it's spottily since 2017 and not at all since 2020. Fortunately, there's a younger generation.

**Comment by OsirisTerje** [comment]
I have added the code from @mxschmitt to the NUnit.Issues repo, here:  https://github.com/nunit/nunit.issues/tree/main/Issue4860.  Thanks @mxschmitt  for allowing that! 

I have also added a series of tests to exclude possible reasons for the issue.  Details in the comments for each test there.

The issue is not related to anything with threading. It must be something that the Kestrel or WebHostBuilder do with respect to its own executioncontext or syncronizationcontext.  The NUnit TestExecutionContext is present, and as one of the tests shows, we can redirect console output from the task running in Kestrel.  It may be related to the way Kestrel (or WebHostBuilder) creates the threads.  Not sure.  But the tests shows that a few things can be eliminated.

The command to run the test for a given test is as follows:

`dotnet test --logger:"console;verbosity=detailed"  --filter "FullyQualifiedName=dotnet_nunit_asp_console_writeline.Tests.TestLoggingInRequestHandler" -- NUnit.DumpXmlTestResults=true`


@CharliePoole  I have also enclosed the Dump file, which has more details than the result file. It is the information as it is received from the engine with the addition of some adapter stuff.
It does **not** contain any of the console output statements.

[E_dotnet-nunit-asp-console-writeline.dll.zip](https://github.com/user-attachments/files/17466438/E_dotnet-nunit-asp-console-writeline.dll.zip)

**Comment by joeldickson** [comment]
I've been looking into this because its been annoying me for a long time now, we use the WebApplcationFactory a LOT, these days im pushing my guys to use more black box testing on our micro-services (RE: spotify testing honey comb) and i want to reduce friction for them.

The problem with this is that NUnit needs to attach the right log output to the write test, in these cases it gets the log output but doesn't know where to route it, so you need to some how lock each instance of kestrel to the right test context, so i've tried to come up with an implementation like this, where you instantiate the factory and it returns a function you can execute in a using that does the reflection attaching thing described in the code issue.

This is how i think it should work for the end user (imo, happy for feedback).

```csharp

    [Test]
    public async Task ExampleTest()
    {
        await using var factory =
            new WebApplicationFactory<Program>()
            .WithNUnitTestLogging(out var scope);

        using (scope)
        {
            var client = factory.CreateClient();
            var response = await client.GetAsync("/");
            response.EnsureSuccessStatusCode();
        }
    }

```

So `WithNUnitTestLogging` is the new implementation that return a IDisposable that in the ctor creates the context, then disposes what we need to at the end of the using.

This way its constructed in the test context and attached properly.

I dont think its a part of NUnit though (happy to be told I'm wrong), its got dependency on Mvc.Testing libraries.

I think i'll put it in another library to start with, but if we think the implementation should live somewhere in NUnit please lmk.

I'll post here again when i'm done with some more details

**Comment by joeldickson** [comment]
I've publish my solution here https://github.com/agoda-com/NUnit.KestrelLogging

**Comment by OsirisTerje** [comment]
@joeldickson Interesting solution, good work!  

Using the WebApplicationFactory is very common, so it is interesting to have a solution to that.

@manfred-brands Can you have a look at this?

**Comment by manfred-brands** [comment]
@OsirisTerje 

In ASP.NET Core the controller runs in its own thread/context. The NUnit test talks to it using Http. 
There is no common thread/async context so it is logical that output is not captured by the tests.

The solution here, captures the `TestExecutionContext` when the `WebApplicationFactory` is created in the nunit test and then sets it in the ASP.NET Core pipeline so it is set when the controller is executed.

This works if the factory is created for every test, either in the test itself as per example code or in a `SetUp` method. 
It doesn't work if the factory is created in a `OneTimeSetUp` method because in NUnit every test gets a new `Outwriter`.

The IntegrationTest at my work creates the factory in the `OneTimeSetUp`, so it wouldn't work there.
